src/types.ts

```diff
@@ -103,55 +107,59 @@ export interface Plugin<
   CtxWithPlugins extends Ctx = Ctx
 > {
   name: string;
-  noClient?: (context: PluginContext<API, Data, G, CtxWithPlugins>) => boolean;
+  noClient?: (
+    context: PluginContext<API, Data, G, Ctx | CtxWithPlugins>
+  ) => boolean;
   setup?: (setupCtx: {
     G: G;
-    ctx: CtxWithPlugins;
-    game: Game<G, CtxWithPlugins>;
+    ctx: Ctx | CtxWithPlugins;
+    game: Game<G, Ctx | CtxWithPlugins>;
   }) => Data;
   action?: (data: Data, payload: ActionShape.Plugin['payload']) => Data;
   api?: (context: {
     G: G;
-    ctx: CtxWithPlugins;
-    game: Game<G, CtxWithPlugins>;
+    ctx: Ctx | CtxWithPlugins;
+    game: Game<G, Ctx | CtxWithPlugins>;
     data: Data;
     playerID?: PlayerID;
   }) => API;
-  flush?: (context: PluginContext<API, Data, G, CtxWithPlugins>) => Data;
+  flush?: (context: PluginContext<API, Data, G, Ctx | CtxWithPlugins>) => Data;
   dangerouslyFlushRawState?: (flushCtx: {
-    state: State<G, CtxWithPlugins>;
-    game: Game<G, CtxWithPlugins>;
+    state: State<G, Ctx | CtxWithPlugins>;
+    game: Game<G, Ctx | CtxWithPlugins>;
     api: API;
     data: Data;
-  }) => State<G, CtxWithPlugins>;
-  fnWrap?: (fn: AnyFn) => (G: G, ctx: CtxWithPlugins, ...args: any[]) => any;
+  }) => State<G, Ctx | CtxWithPlugins>;
+  fnWrap?: (
+    fn: AnyFn
+  ) => (G: G, ctx: Ctx | CtxWithPlugins, ...args: any[]) => any;
 }
 
 type MoveFn<G extends any = any, CtxWithPlugins extends Ctx = Ctx> = (
   G: G,
-  ctx: CtxWithPlugins,
+  ctx: Ctx | CtxWithPlugins,
```

Comment:
Conversely, moves and hooks _will_ have the `CtxWithPlugins`. (This is effectively the plain and “enhanced” contexts mentioned previously.)

Reply:
(Hope this all makes sense!)

Reply:
Yes, that issue you linked was very insightful. I've updated my PR

Reply:
Thanks Crhistian! What I meant here was that for moves and hooks we can actually type them as `(G: G, ctx: CtxWithPlugins) => any`, because the context will always be enhanced before they are called. Sorry not to be clearer. I can go through and tidy up the places where we can use `CtxWithPlugins` instead of `Ctx | CtxWithPlugins` if necessary.

Reply:
Ohhhhhh got it. Yeah, I can update it no problem

Reply:
Awesome, thanks for hanging in there!

Reply:
@delucis so if I understand correctly I think I want to revert my changes regarding `Ctx | CtxWithPlugins` and make just `State` be `ctx: Ctx | CtxWithPlugins` and have everything else be `ctx: CtxWithPlugins` does that seem right?

If not, I might need an assist. Not sure if I have a good enough understanding of the library yet to make the subtle distinctions

Reply:
Yes, after a quick look through, I think that’s correct. The only one I’m not 100% on is the `processMove` method. I’ll have a look at the source for that.

Reply:
Looks like based on the source code `processMove` should also have `CtxWithPlugins` so I updated my PR as such

```
// src/core/reducer.ts

// Execute plugins.
state = plugins.Enhance(state, {
  game,
  isClient,
  playerID: action.payload.playerID,
});

// Process the move.
let G = game.processMove(state, action.payload);
```